Sources/Workspace/WorkspaceConfiguration.swift

```diff
@@ -209,6 +209,107 @@ extension Workspace {
+                providers.append(try NetrcAuthorizationProvider(path: path, fileSystem: localFileSystem))
+            case .`default`(let rootPath):
+                // package/project "local" .netrc file, takes priority over user-level file
+                let localPath = rootPath.appending(component: ".netrc")
+                // user didn't tell us to explicitly use these .netrc files so be more lenient with errors
+                if let localProvider = self.tryLoadNetrc(path: localPath, fileSystem: fileSystem, observabilityScope: observabilityScope) {
+                    providers.append(localProvider)
+                }
+
+                // user .netrc file (most typical)
+                let userHomePath = fileSystem.homeDirectory.appending(component: ".netrc")
+                // user didn't tell us to explicitly use these .netrc files so be more lenient with errors
+                if let userHomeProvider = self.tryLoadNetrc(path: userHomePath, fileSystem: fileSystem, observabilityScope: observabilityScope) {
+                    providers.append(userHomeProvider)
+                }
+            case .disabled:
+                // noop
+                break
+            }
+
+            switch self.configuration.keychain {
+            case .enabled:
+                #if canImport(Security)
+                providers.append(KeychainAuthorizationProvider(observabilityScope: observabilityScope))
+                #else
+                assert("Keychain not supported on this platform")
+                #endif
+            case .disabled:
+                // noop
+                break
+            }
+
+            return providers.isEmpty ? .none : CompositeAuthorizationProvider(providers, observabilityScope: observabilityScope)
+        }
+
+        private func tryLoadNetrc(
+            path: AbsolutePath,
+            fileSystem: FileSystem,
+            observabilityScope: ObservabilityScope
+        ) -> NetrcAuthorizationProvider? {
+            guard fileSystem.exists(path) else {
+                return .none
+            }
+
+            do {
+                return try NetrcAuthorizationProvider(path: path, fileSystem: fileSystem)
+            } catch {
+                observabilityScope.emit(warning: "Failed to load .netrc file at \(path). Error: \(error)")
+                return .none
+            }
+        }
+
+        public struct Configuration {
+            public var netrc: Netrc
+            public var keychain: Keychain
+
+            public init (netrc: Netrc, keychain: Keychain) {
+                self.netrc = netrc
+                self.keychain = keychain
+            }
+
+            public init () {
+                self.init(
+                    netrc: .disabled,
+                    keychain: .disabled
+                )
+            }
+
+            public enum Netrc {
+                case disabled
+                case custom(AbsolutePath)
+                case `default`(rootPath: AbsolutePath)
```

Comment:
Would it make sense to make `rootPath` optional here or split this up into two cases? I see the case for "default global" and "default global plus local at this root path" (this is what we are doing right now).

Reply:
yes had the same thought and changed accordingly